plutus-emulator/src/Wallet/Emulator/NodeClient.hs

```diff
@@ -0,0 +1,150 @@
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE DeriveAnyClass        #-}
+{-# LANGUAGE DerivingStrategies    #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE TupleSections         #-}
+{-# LANGUAGE TypeApplications      #-}
+module Wallet.Emulator.NodeClient where
+
+import           Conduit
+import           Control.Lens                 hiding (index)
+import           Control.Monad.Except
+import           Control.Monad.State          
+import           Data.Aeson                   (FromJSON, ToJSON)
+import           Data.List                    (partition)
+import           Data.Maybe                   (isNothing)
+import           Data.Traversable             (for)
+import           GHC.Generics                 (Generic)
+
+import           Ledger                       (Block, Blockchain, Slot (..), Tx (..), TxId, hashTx)
+import qualified Ledger.Index                 as Index
+import qualified Ledger.Interval              as Interval
+
+
+-- | Events produced by the blockchain emulator.
+data NodeClientEvent =
```

Comment:
Similarly, I think these are *chain* events. 

I think we might want more modules.
- `Chain.Emulator`, which has the stuff you need to emulate the global evolution of the chain. That is: the state of such an emulator (`ChainState`), and the events that represent changes to it (`ChainEvent`). Perhaps also the functions that mutate it.
- `NodeClient`, which has the actual interface to the node client.
- `NodeClient.Emulator` (?) which has an implementation of the client interface backed by the emulation.

You probably don't need to split stuff up now, but what do you think about it conceptually?